### ISSUE TITLE
Optimize the performance on a large amount of code

### ISSUE THREAD

**Issue by yescorp** [akvelon/flutter-code-editor#170]

**Comment by yescorp** [comment]
[Report](https://drive.google.com/file/d/1oxLrXlzt5OqXShFBT4ynQqctq-PhkScz/view?usp=sharing)
Before changes
[With 485 lines of code](https://drive.google.com/file/d/1RdZN6bYNQBMhrIGz-JCD5d8Sfm7oIKrz/view?usp=sharing).
Mode: Debug
Analysis of a single KeyPressEvent

- KeyPressEvent `128.64ms`
  - Whole setValue `119.17ms`
    - CodeController methods `32.94ms`
      - Parse highlight `12.92ms` + `1.15ms` + `4.24ms` = `18.31ms`
      - Process single line comments `2.69ms`
      - Text to code lines `1.09ms`
      - Comments to hidden ranges `0.62ms`
      - cutHighlighted `1.32ms`
      - split lines `2.54ms`
      - folded as (with no folded blocks) `5.85ms`
    - super.setValue `46.42ms`
      - layout `42.29ms`
      - textPainter `39.16ms`
    - setText `38.54ms`
      - Autocompleter.autocomplete `38.54ms`
        - enterList `29.66ms` + `6.47ms` = `36.13ms`
- Animation frame fired `509.50ms`

<img width="657" alt="Screen Shot 2023-01-31 at 12 31 24" src="https://user-images.githubusercontent.com/79095619/215684110-cf753bbd-9d64-447b-9ce2-22d0ca0ab5b8.png">
<img width="1051" alt="Screen Shot 2023-01-31 at 12 32 20" src="https://user-images.githubusercontent.com/79095619/215684268-136847a7-5b14-4f57-a495-723d922f91b3.png">

**Comment by yescorp** [comment]
If there is only 1 TextSpan to paint:

<img width="446" alt="Screen Shot 2023-01-31 at 18 45 09" src="https://user-images.githubusercontent.com/79095619/215763796-b98b36bc-ebc0-4c50-84fd-00615b02c651.png">

**Comment by yescorp** [comment]
As I tested, even if `TextField` uses its own scroll, it doesn't optimize the underlying painting mechanism.
And the whole content of it is repainted each time.

Optimizing `TextSpan` generation is also useless, because `TextField` doesn't cache anything and even if receives the exact same `TextSpan` object by reference, it still repaints everything.

Using raw `TextField` widget with its own scroll controls (485 lines of code):
<img width="441" alt="Screen Shot 2023-01-31 at 19 04 31" src="https://user-images.githubusercontent.com/79095619/215767547-6fedba6d-5a0d-4152-94af-5eb1416a9381.png">

Using raw `TextField` with 0 symbols in it:
<img width="485" alt="Screen Shot 2023-01-31 at 19 05 23" src="https://user-images.githubusercontent.com/79095619/215767738-c1cba9e5-7bc9-44d1-93b7-59c7a177b89c.png">

**Comment by yescorp** [comment]
@alexeyinkin I think your idea with joining `TextSpan`s in invisible areas of the CodeField might work. 
I am going to try to implement it now.

**Comment by yescorp** [comment]
Addition to the previous comment:
Yes, it does take a little time to paint, but only if the `TextPainter.layout` was called.
But, when we change the code, we need to call that method each time to re-calculate the layout.
(`TextPainter` is used inside `TextField` of course)

Drawing - cheap.
Layout - expensive.

I looked into the source code of `TextPainter` a little, and find out:
1. layout of the text is performed natively, with the underlying class `Paragraph`
2. `Paragraph` is then being painted onto the canvas inside `TextPainter` with the method `drawParagraph`.

My thoughts about this:
1. Cache paragraphs.
- So that there is no need to call layout on them?
  For example we can store paragraphs for:
  - keywords (fixed amount)
  - letters with different styling.
    - 36 regular 
    - 36 with comment color
    - 36 with quote color

2. Use different TextPainter for each line.
- So that when we update the code we will calculate the layout only for changed lines.
- And repaint changed lines.

For both cases
- We can calculate the whole layout of the `CodeField` by using line-height of a single letter and allow it to scale infinitely horizontally. Therefore, we will not need to call layout on the whole text.

<img width="951" alt="Screen Shot 2023-02-05 at 14 31 03" src="https://user-images.githubusercontent.com/79095619/216809202-7c1bbcde-400e-4c92-ba3a-085de6656913.png">
<img width="911" alt="Screen Shot 2023-02-05 at 14 31 18" src="https://user-images.githubusercontent.com/79095619/216809218-49bf4378-d570-48c3-80c3-a8afa0a09089.png">

**Comment by yescorp** [comment]
If I am not mistaken, there is no way to optimize the layout that is called inside of a `TextField`.

But, implementing the painting from scratch will require a lot of additional work:
1. Draw Toggling Cursor
2. Draw Selection
3. Handle mouse events (tap, double tap, select lines, etc)

Before deciding whether we need to implement custom painting, we need to calculate the estimated increase in performance first. I am currently working on that.

**Comment by yescorp** [comment]
Rendering the following Text:
`import package return () {} /*   */ // comment`

- Layout  ~2 ms 
- Paint  ~10 ms
<img width="512" alt="Screen Shot 2023-02-07 at 17 38 51" src="https://user-images.githubusercontent.com/79095619/217235050-67092665-0f97-4631-8695-bcb318660cd8.png">
<img width="505" alt="Screen Shot 2023-02-07 at 17 38 40" src="https://user-images.githubusercontent.com/79095619/217235069-27418da2-031c-4105-9f05-2c0aa5b75b7f.png">


If we have 500 lines of code we will need to call paint for each of them, which will slow down to:
- Layout  ~ 10 ms
- Paint ~70 ms

<img width="750" alt="Screen Shot 2023-02-07 at 17 35 35" src="https://user-images.githubusercontent.com/79095619/217234547-d5ec01f5-459e-4048-824c-daa871f6a0f6.png">
<img width="847" alt="Screen Shot 2023-02-07 at 17 35 27" src="https://user-images.githubusercontent.com/79095619/217234694-e95d4319-5bc4-4c77-8229-432c7dcd4b6e.png">

In total we have:
Increase in load = 500 times.
Slow down = 8 times

Of course in the reality we will probably have worse results.

**Comment by yescorp** [comment]
Here is the suggestion from ChatGPT:

Here's an example of a custom widget in Flutter that uses CustomPaint to display text on the screen and provides an opportunity to edit it:

<details>
  <summary>Code</summary>

```dart
import 'package:flutter/material.dart';

class CustomTextPainter extends CustomPainter {
  final String text;
  final TextEditingController controller;
  final FocusNode focusNode;

  CustomTextPainter({
    @required this.text,
    @required this.controller,
    @required this.focusNode,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final textSpan = TextSpan(
      text: text,
      style: TextStyle(
        fontSize: 32.0,
        color: Colors.black,
      ),
    );
    final textPainter = TextPainter(
      text: textSpan,
      textDirection: TextDirection.ltr,
    );
    textPainter.layout(
      minWidth: 0,
      maxWidth: size.width,
    );
    textPainter.paint(canvas, Offset(0, 0));
  }

  @override
  bool shouldRepaint(CustomTextPainter oldDelegate) {
    return oldDelegate.text != text;
  }
}

class TextEditor extends StatefulWidget {
  final String text;

  TextEditor({
    @required this.text,
  });

  @override
  _TextEditorState createState() => _TextEditorState();
}

class _TextEditorState extends State<TextEditor> {
  final _controller = TextEditingController();
  final _focusNode = FocusNode();

  @override
  void initState() {
    _controller.text = widget.text;
    super.initState();
  }

  @override
  void dispose() {
    _controller.dispose();
    _focusNode.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: 200,
      width: double.infinity,
      child: Stack(
        children: [
          CustomPaint(
            painter: CustomTextPainter(
              text: _controller.text,
              controller: _controller,
              focusNode: _focusNode,
            ),
          ),
          Positioned.fill(
            child: TextField(
              controller: _controller,
              focusNode: _focusNode,
              style: TextStyle(
                fontSize: 32.0,
                color: Colors.transparent,
              ),
              decoration: InputDecoration(
                border: InputBorder.none,
              ),
              onTap: () {
                FocusScope.of(context).requestFocus(_focusNode);
              },
            ),
          ),
        ],
      ),
    );
  }
}
```
</details>

In this example, we have a custom painter CustomTextPainter that takes a text parameter, a `TextEditingController` and a `FocusNode` to manage the text input. This painter is used by the `TextEditor` widget to render the text on the screen, which is editable through a `TextField` widget positioned on top of the `CustomPaint`

**Comment by yescorp** [comment]
With the 485 lines code.
ChatGPT inspired me to the idea of delegating all of the handles to the TextField while drawing all of the text in the background separately.

However..

Rendering text inside a TextField with TextEditingController is still slow..

Considering that we will also need to draw the text under it.
This solution doesn't feel to be good.

<img width="613" alt="Screen Shot 2023-02-07 at 18 22 46" src="https://user-images.githubusercontent.com/79095619/217243800-73c9e5cf-9143-4c6a-96bc-e8013dbf2b77.png">

**Comment by yescorp** [comment]
I implemented small TextField where the text can only be added to the end.
Test results with the same code snippet:

- Improvement in paint: 10 times (without painting cursor and selection).

Now, the problem is optimizing TextSpan Generation.

<img width="1006" alt="Screen Shot 2023-02-10 at 12 56 13" src="https://user-images.githubusercontent.com/79095619/218024160-28a68680-a0b0-4765-9ab5-a09cffa242da.png">
<img width="948" alt="Screen Shot 2023-02-10 at 12 56 21" src="https://user-images.githubusercontent.com/79095619/218024169-f3e376ef-249c-41f8-a1a3-dbe069ab533e.png">
<img width="683" alt="Screen Shot 2023-02-10 at 12 56 35" src="https://user-images.githubusercontent.com/79095619/218024172-5fc045c3-a656-4e07-a5aa-944cd5d57d02.png">

**Comment by yescorp** [comment]
UPD:
Turns out I was caching the TextPainters wrong in the last comment.
Results with properly cached TextPainters:

<img width="397" alt="Screen Shot 2023-02-11 at 00 06 08" src="https://user-images.githubusercontent.com/79095619/218164624-f079f405-07fc-42be-af9c-746fb2b6b8e4.png">

<img width="460" alt="Screen Shot 2023-02-11 at 00 07 56" src="https://user-images.githubusercontent.com/79095619/218164880-0da2ea0d-0bf3-4d47-a4f3-395ebb68c88c.png">

**Comment by yescorp** [comment]
NEW PROBLEM:
The width of the letters are not the same. 
Trying to find anything that can transform the text to a monospace..

**Comment by yescorp** [comment]
It is getting much more complicated.
Caching the `TextPainters` for `TextSpans` works fine until some threshold where it becomes a little laggy again.

Caching `TextPainters` for individual letters also gets slow as the amount of text increases.

I also tried using `ListView.builder` and wrap each line of code into animated builder. But when the line count changes, the ListView is being rebuilt entirely.
I tried to pre-create 10K code lines in ListView, but when it is build for the first time it actually builds all of the elements, so building such list takes like 5 minutes =)

Or maybe I am using `ListView.builder` wrongly, I am not so sure about how it works and how to increase its performance.

**Comment by yescorp** [comment]
I think I finally found the solution we need.
If we use `CustomScrollView` with Slivers inside, we will have the opportunity to build only those line, that are currently visible on the screen. I was able to reach the best performance in terms of updating the text by using this approach.

The following is done by just painting each line with `RichText` and updating the controller by using `Focus`.
<img width="941" alt="Screen Shot 2023-02-17 at 19 04 57" src="https://user-images.githubusercontent.com/79095619/219660794-8c8726f3-cfe6-4889-a70e-2b64afee7adb.png">
